packages/eslint/src/plugins/plugin.ts

```diff
@@ -28,48 +29,64 @@ export const createNodes: CreateNodes<EslintPluginOptions> = [
     (context as any).configFiles = context.configFiles ?? [];
 
-    // Create a Set of all the directories containing eslint configs
-    const eslintRoots = new Set(context.configFiles.map(dirname));
-    const configDir = dirname(configFilePath);
+    // Create a Set of all the directories containing eslint configs, and a
+    // list of globs to exclude from child projects
+    const eslintRoots = new Set();
+    const nestedEslintRootPatterns: string[] = [];
+    for (const configFile of context.configFiles) {
+      const eslintRootDir = dirname(configFile);
+      eslintRoots.add(eslintRootDir);
+
+      if (eslintRootDir !== configDir && isSubDir(configDir, eslintRootDir)) {
+        nestedEslintRootPatterns.push(`${eslintRootDir}/**/*`);
+      }
+    }
 
-    const childProjectRoots = globWithWorkspaceContext(
+    const childProjectRoots = new Set<string>();
+    const ESLint = resolveESLintClass(isFlatConfig(configFilePath));
+
+    const projectFiles = globWithWorkspaceContext(
       context.workspaceRoot,
       [
         'project.json',
         'package.json',
         '**/project.json',
         '**/package.json',
-      ].map((f) => join(configDir, f))
-    )
-      .map((f) => dirname(f))
-      .filter((childProjectRoot) => {
-        // Filter out projects under other eslint configs
-        let root = childProjectRoot;
-        // Traverse up from the childProjectRoot to either the workspaceRoot or the dir of this config file
-        while (root !== dirname(root) && root !== dirname(configFilePath)) {
-          if (eslintRoots.has(root)) {
-            return false;
-          }
-          root = dirname(root);
-        }
-        return true;
-      })
-      .filter((dir) => {
-        // Ignore project roots where the project does not contain any lintable files
-        const lintableFiles = globWithWorkspaceContext(context.workspaceRoot, [
-          join(dir, `**/*.{${options.extensions.join(',')}}`),
-        ]);
-        return lintableFiles.length > 0;
+      ].map((f) => join(configDir, f)),
+      nestedEslintRootPatterns.length ? nestedEslintRootPatterns : undefined
+    );
+    for (const projectFile of projectFiles) {
+      const childProjectRoot = dirname(projectFile);
+      if (childProjectRoots.has(childProjectRoot)) {
+        continue;
+      }
+
+      // Ignore project roots where the project does not contain any lintable files
+      const lintableFiles = globWithWorkspaceContext(
+        context.workspaceRoot,
+        [join(childProjectRoot, `**/*.{${options.extensions.join(',')}}`)],
+        nestedEslintRootPatterns.length ? nestedEslintRootPatterns : undefined
+      );
+      const eslint = new ESLint({
+        cwd: join(context.workspaceRoot, childProjectRoot),
       });
+      for (const file of lintableFiles) {
+        if (!(await eslint.isPathIgnored(join(context.workspaceRoot, file)))) {
+          childProjectRoots.add(childProjectRoot);
+          break;
+        }
+      }
```

Comment:
We should probably use a `Promise.all` here. But you can address it in a followup PR.